src/python/pants/backend/experimental/visibility/register.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.internals.dep_rules import (
+    BuildFileDependencyRules,
+    BuildFileDependencyRulesImplementation,
+    BuildFileDependencyRulesImplementationRequest,
+)
+from pants.engine.internals.target_adaptor import TargetAdaptor, TargetAdaptorRequest
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    FieldSet,
+    ValidatedDependencies,
+    ValidateDependenciesRequest,
+)
+from pants.engine.unions import UnionRule
+
+
+class BuildFileVisibilityImplementationRequest(BuildFileDependencyRulesImplementationRequest):
+    pass
+
+
+class VisibilityValidateFieldSet(FieldSet):
+    required_fields = (Dependencies,)
+
+
+class VisibilityValidateDependenciesRequest(ValidateDependenciesRequest):
+    field_set_type = VisibilityValidateFieldSet
+
+
+@rule
+def build_file_visibility_implementation(
+    _: BuildFileVisibilityImplementationRequest,
+) -> BuildFileDependencyRulesImplementation:
+    return BuildFileDependencyRulesImplementation(BuildFileDependencyRules)
+
+
+@rule
+async def visibility_validate_dependencies(
+    request: VisibilityValidateDependenciesRequest,
+) -> ValidatedDependencies:
+    address = request.field_set.address.maybe_convert_to_target_generator()
+    _ = await MultiGet(
+        Get(
+            TargetAdaptor,
+            TargetAdaptorRequest(
```

Comment:
So, If I wanted to get the dependency rules from inside the python inference system, would I do a `Get(TargetAdapter, TargetAdapterRequest(...))` like this?

Reply:
If you include `address_of_origin`, you will either get the requested target adaptor (if there's no rule `deny`ing the dependency between the two) or an error will be raised.

As you noted in the other thread, getting the low level `AddressFamily` allows you to interact with the rules directly.

Reply:
Reworked the API so you can now request the rule for each dependency for a given target.